api/test/wfl/integration/datarepo_test.clj

```diff
@@ -77,20 +78,86 @@
+         (is (= % (:id snapshot)))))))
+
+(deftest test-flattened-query-result
+  (let [samplesheets (-> (datarepo/dataset testing-dataset)
+                         (datarepo/query-table "flowcell" [:samplesheets])
+                         :rows
+                         (->> (mapcat first)))]
+    (is (every? string? samplesheets) "Nested arrays were not normalized.")))
+
+(defn ^:private table->map-view
+  "Create a hash-map 'view' of the BigQuery `table`. A 'view' is one"
+  [{:keys [schema rows] :as _table}]
+  (let [make-entry (fn [idx field] [(-> field :name keyword) idx])
+        key->index (into {} (map-indexed make-entry (:fields schema)))]
+    (map (util/curry #(when-let [idx (key->index %2)] (%1 idx))) rows)))
+
+(defn ^:private maps->table
+  "Transform a list of maps into a table with columns given by `attributes`."
+  [maps attributes]
+  (let [table {:schema {:fields (mapv #(-> {:name (name %)}) attributes)}}
+        f     #(reduce (fn [row attr] (conj row (% attr))) [] attributes)]
+    (assoc table :rows (map f maps))))
+
+(defn ^:private make-entity-import-request-tsv
+  [from-dataset primary-key maps]
+  ;; the columns need to be ordered such that that primary key comes first
+  (let [columns (>>> #(dissoc % primary-key) keys sort #(cons primary-key %))]
+    (-> (map #(datasets/rename-gather % from-dataset) maps)
+        (maps->table (columns from-dataset))
+        bigquery/dump-table->tsv)))
+
+(defn import-table
+  "Import the BigQuery `table` into the `entity` in the Terra `workspace`,
+   using `from-snapshot` to map column names in `snapshot` to the names in the
+   workspace `table`. Return `[entity name]` pairs of the entities imported
+   into the workspace."
+  [table workspace entity primary-key from-snapshot]
+  (let [maps (table->map-view table)]
+    (->> (make-entity-import-request-tsv from-snapshot primary-key maps)
+         .getBytes
+         (firecloud/import-entities workspace))
+    (map (comp #(-> [entity %]) #(% primary-key)) maps)))
+
+;; not sure where this should live!
+(def ^:private snapshot-id "7cb392d8-949b-419d-b40b-d039617d2fc7")
+(deftest test-import-snapshot
+  (let [dataset-table "flowcell"
+        entity        "flowcell"
+        entity-key    :flowcell_id
+        from-dataset  (resources/read-resource "entity-from-dataset.edn")]
+    (fixtures/with-temporary-workspace
+      (fn [workspace]
+        (let [entities
+              (-> (datarepo/snapshot snapshot-id)
+                  (datarepo/query-table dataset-table)
+                  (import-table workspace entity entity-key from-dataset))]
+          (try
+            (let [names (->> #(firecloud/list-entities workspace entity)
+                             (util/poll-while empty?)
+                             (map :name)
+                             set)]
+              (doseq [[_ name] entities] (is (contains? names name))))
+            (finally (firecloud/delete-entities workspace entities))))))))
```

Comment:
This could `(keep :name)` and drop `contains?`.

Reply:
Wouldn't `(keep :name)` and `(map :name)` both deliver the same thing?